engine/src/main/java/org/apache/apex/log/appender/ApexRollingFileAppender.java

```diff
@@ -0,0 +1,174 @@
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+
+import org.apache.log4j.RollingFileAppender;
+import org.apache.log4j.helpers.CountingQuietWriter;
+import org.apache.log4j.helpers.LogLog;
+import org.apache.log4j.spi.LoggingEvent;
+
+/**
+ * ApexRollingFileAppender extends {@link RollingFileAppender}
+ * This appender doesn't keep renaming files when file reach max size.
+ * The appender starts with <filename>.0 file when file reaches max size,
+ * it creates next file <filename>.1 and so on till <filename>.MaxBackupIndex.
+ * After that it will delete <filename>.0 and start creating log files in that
+ * order again.
+ * To identify current file, it creates <filename> symlink to current file.
+ */
+public class ApexRollingFileAppender extends RollingFileAppender
+{
+  private int currentFileIndex = 0;
+  private String currentFileName = "";
+  private long lastExceptionOffset;
+
+  public synchronized void setFile(String fileName, boolean append, boolean bufferedIO, int bufferSize)
+      throws IOException
+  {
+    currentFileName = fileName + "." + currentFileIndex;
+    LogLog.debug("setFile called: " + currentFileName + ", " + append);
+
+    // It does not make sense to have immediate flush and bufferedIO.
+    if (bufferedIO) {
+      setImmediateFlush(false);
+    }
+
+    reset();
+    FileOutputStream ostream = null;
+    try {
+      //   attempt to create file
+      ostream = new FileOutputStream(currentFileName, append);
+    } catch (FileNotFoundException ex) {
+      //   if parent directory does not exist then
+      //      attempt to create it and try to create file
+      //      see bug 9150
+      String parentName = new File(currentFileName).getParent();
+      if (parentName != null) {
+        File parentDir = new File(parentName);
+        if (!parentDir.exists() && parentDir.mkdirs()) {
+          ostream = new FileOutputStream(currentFileName, append);
+        } else {
+          throw ex;
+        }
+      } else {
+        throw ex;
+      }
+    }
+    Writer fw = createWriter(ostream);
+    if (bufferedIO) {
+      fw = new BufferedWriter(fw, bufferSize);
+    }
+
+    if (new File(fileName).exists()) {
+      Files.delete(Paths.get(fileName));
+    }
+    Files.createSymbolicLink(Paths.get(fileName), Paths.get(currentFileName));
+    this.setQWForFiles(fw);
+    this.fileName = fileName;
+    this.fileAppend = append;
+    this.bufferedIO = bufferedIO;
+    this.bufferSize = bufferSize;
+    writeHeader();
+    LogLog.debug("setFile ended");
+  }
+
+  @Override
+  public void rollOver()
+  {
+    if (qw != null) {
+      long size = ((CountingQuietWriter)qw).getCount();
+      LogLog.debug("rolling over count=" + size);
+    }
+    LogLog.debug("maxBackupIndex=" + maxBackupIndex);
+
+    boolean deleteSucceeded = true;
+    int nextFileIndex = (currentFileIndex + 1) % maxBackupIndex;
+    String nextFileName = fileName + '.' + nextFileIndex;
```

Comment:
Yes this is confusing, I had raised that point at time of proposing this design. And that's why we though we would have a soft link to point to current file. If we wish you have fixed log file names we have to change logger behavior to something, and provided that something is different from standard log rotation, this is bound to be confusing.